<!doctype html><html><head><title> ACE/ | Barton lab</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" href="/assets/css/main.css"><link rel="stylesheet" href="/assets/css/syntax.css"><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Avenir"><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Code+Pro"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css"> <script src="/assets/js/katex.min.js"></script><link href="/assets/css/katex.min.css" rel="stylesheet" type="text/css"> <script src="/assets/js/auto-render.min.js"></script> <script>!function(e,a,t,n,c,o,s){e.GoogleAnalyticsObject=c,e[c]=e[c]||function(){(e[c].q=e[c].q||[]).push(arguments)},e[c].l=1*new Date,o=a.createElement(t),s=a.getElementsByTagName(t)[0],o.async=1,o.src=n,s.parentNode.insertBefore(o,s)}(window,document,"script","//www.google-analytics.com/analytics.js","ga"),ga("create","UA-89355185-1","auto"),ga("send","pageview")</script><title>ACE/ | Barton lab at UCR Physics</title><meta property="og:title" content="ACE/"/><meta name="author" content="John P Barton"/><meta property="og:locale" content="en_US"/><meta name="description" content="Research in computational immunology, host-pathogen dynamics, and statistical inference"/><meta property="og:description" content="Research in computational immunology, host-pathogen dynamics, and statistical inference"/><link rel="canonical" href="http://localhost:4000/projects/ACE/"/><meta property="og:url" content="http://localhost:4000/projects/ACE/"/><meta property="og:site_name" content="Barton lab at UCR Physics"/><meta name="twitter:card" content="summary"/><meta name="twitter:site" content="@_jpbarton"/><meta name="twitter:creator" content="@John P Barton"/> <script type="application/ld+json">
{"name":null,"description":"Research in computational immunology, host-pathogen dynamics, and statistical inference","author":{"@type":"Person","name":"John P Barton"},"@type":"WebPage","url":"http://localhost:4000/projects/ACE/","publisher":null,"image":null,"headline":"ACE/","dateModified":null,"datePublished":null,"sameAs":null,"mainEntityOfPage":null,"@context":"http://schema.org"}</script><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="Barton lab at UCR Physics"/></head><body><div class="container"><header class="masthead"><h3 class="masthead-title"> <a href="/menu/about.html" class="masthead-lab">Barton lab</a> <small class="masthead-subtitle">UCR Physics</small><div class="menu"> <nav class="menu-content"> <a href="/menu/../index.html">Posts</a> <a href="/menu/research.html">Research</a> <a href="/menu/papers.html">Papers</a> <a href="/menu/code.html">Code</a> <a href="/menu/members.html">Members</a> <a href="/menu/join.html">Join</a> </nav> <nav class="social-icons"> <a href="https://www.github.com/bartonlab" target="_blank"><i class="fa fa-github" aria-hidden="true"></i></a> <a href="https://scholar.google.com/citations?user=ItAcAOMAAAAJ" target="_blank"><i class="fa fa-book" aria-hidden="true"></i></a> <a href="https://twitter.com/_jpbarton" target="_blank"><i class="fa fa-twitter" aria-hidden="true"></i></a> <a href="mailto:john.barton@ucr.edu" target="_blank"><i class="fa fa-envelope" aria-hidden="true"></i></a> </nav></div></h3> </header><div class="post-container"><h1><div class="breadcrumb"> ACE</div></h1><p><h3> Fast and flexible code for solving the inverse Ising/Potts inference problem</h3><div class="bigspacer"></div><div class="grid"><div class="project-side"><div class="bigspacer"></div> <small> Source code<div class="smallspacer"></div> <i class="fa fa-cog fa-fw"></i> <a href="https://github.com/johnbarton/ACE">ACE</a> </small><div class="bigspacer"></div> <small> Contributors<div class="smallspacer"></div> <img class="github-avatar" src="https://avatars0.githubusercontent.com/u/2357804?v=4&s=50"> <a class="github-handle" href="https://github.com/johnbarton"> johnbarton </a> </small><div class="bigspacer"></div> <small> Latest commits<ul class="commit-list"><div class="smallspacer"></div><li class="commit-list"> <i class="fa fa-check-square-o fa-fw"></i> <a href="https://github.com/johnbarton/ACE/commit/d73eaee75fed52f9f42b6e321f20bfef3a6c34ae"> <span class="commit-date">13 Sep 2017</span><span class="commit-message">- more informative error messages when input files do not exist/can't be accessed</span> </a></li><div class="smallspacer"></div><li class="commit-list"> <i class="fa fa-check-square-o fa-fw"></i> <a href="https://github.com/johnbarton/ACE/commit/27ee6a740a49689994a0d12ab92a644e89d255ce"> <span class="commit-date">13 Sep 2017</span><span class="commit-message">- compatibility fix</span> </a></li><div class="smallspacer"></div><li class="commit-list"> <i class="fa fa-check-square-o fa-fw"></i> <a href="https://github.com/johnbarton/ACE/commit/a15c1523db1b47c55d7c012ed055bc242e3511bd"> <span class="commit-date">12 Sep 2017</span><span class="commit-message">- added cmsa output for binary input with Python script</span> </a></li><div class="smallspacer"></div><li class="commit-list"> <i class="fa fa-check-square-o fa-fw"></i> <a href="https://github.com/johnbarton/ACE/commit/c8f80b74e57942c74c826cf80eecb2175c3581b6"> <span class="commit-date">28 Aug 2017</span><span class="commit-message">- fixed error computing pair correlations when using non-consensus gauge</span> </a></li><div class="smallspacer"></div><li class="commit-list"> <i class="fa fa-check-square-o fa-fw"></i> <a href="https://github.com/johnbarton/ACE/commit/4818775c6f7ba8152a70dc4fde3721d69789ab62"> <span class="commit-date">09 Jul 2017</span><span class="commit-message">- Published with https://stackedit.io/</span> </a></li></ul> </small><div class="spacer"></div></div><div class="project-main"><div class="post" id="post"><p> <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><h4 id="ace--">ACE <a href="https://travis-ci.org/johnbarton/ACE"><img src="https://travis-ci.org/johnbarton/ACE.svg?branch=master" alt="Build Status"/></a> <a href="https://zenodo.org/badge/latestdoi/32744304"><img src="https://zenodo.org/badge/32744304.svg" alt="DOI"/></a></h4><h1 id="table-of-contents">Table of contents</h1><p>[TOC]</p><h1 id="introduction">Introduction</h1><p>ACE is a software package designed to quickly and accurately infer <a href="https://en.wikipedia.org/wiki/Ising_model">Ising</a> or <a href="https://en.wikipedia.org/wiki/Potts_model">Potts</a> models based on correlation data from a variety of biological and artificial systems. This software makes use of the <b>A</b>daptive <b>C</b>luster <b>E</b>xpansion (ACE) algorithm.</p><p>Given a set of correlation data or sequence input in <a href="http://en.wikipedia.org/wiki/FASTA_format">FASTA</a> format, ACE will produce a Ising or Potts model that reproduces the input correlations to within the expected error due to finite sampling.</p><p><b>NOTE:</b> Mathematical expressions through MathJax are currently not supported on GitHub. To see these expressions rendered properly, please see <code class="highlighter-rouge">README.pdf</code> or the alternative <a href="http://johnbarton.github.io/projects/ACE/">project page</a>.</p><h1 id="installation">Installation</h1><p>Download and unzip the package, then run the following commands in the terminal from the new directory:</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>./configure
<span class="nv">$ </span>make
</code></pre></div></div><p>If you’d like to be able to run the program from any directory, you can then enter:</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>make install
</code></pre></div></div><h1 id="required-input">Required Input</h1><p>Running the algorithm requires a set of correlations as input, to be computed from your data.</p><p>As an example, let’s consider a system of $N$ variables described by the configuration $\underline{x}={x_1, x_2,\ldots,x_N}$, with each variable $x_i$ taking one of $q_i$ possible values, $x_i\in{1,2,\ldots,q_i}$. From a set of $B$ observations of the system, we can compute the frequency of each variable as well as the pairwise correlations,</p><script type="math/tex; mode=display">% <![CDATA[
\begin{aligned}
\begin{align}
p_i(a) &= \frac{1}{B}\sum_{k=1}^{B}\delta(x_i,a)\,,\\
p_{ij}(a,b) &= \frac{1}{B}\sum_{k=1}^{B}\delta(x_i,a)\delta(x_j,b)\,.
\end{align}
\end{aligned} %]]></script><p>Here $\delta$ represents the <a href="http://en.wikipedia.org/wiki/Kronecker_delta">Kronecker delta function</a>. These correlations should be saved in a file ending with the extension <code class="highlighter-rouge">.p</code>, in the following format:</p><blockquote><p>$p_1(1)$ $p_1(2)$ … $p_1(q_1-1)$ $p_2(1)$ $p_2(2)$ … $p_2(q_2-1)$ … $p_N(1)$ $p_N(2)$ … $p_N(q_N-1)$ $p_{1,2}(1,1)$ $p_{1,2}(1,2)$ … $p_{1,2}(1,q_2-1)$ $p_{1,2}(2,1)$ $p_{1,2}(2,2)$ … $p_{1,2}(q_1-1,q_2-1)$ $p_{1,3}(1,1)$ …</p></blockquote><p>In other words, the first $N$ lines of the file record the frequency that each state is observed at each site, and the next $N(N-1)/2$ lines record the pairwise correlations. Note that, because $\sum_{a=1}^{q_i} p_i(a)=1$, the frequency (and corresponding pair correlations) for one state at each site need not be specified explicitly.</p><p>These values should be given in floating point or scientific format, with whitespace (e.g. <code class="highlighter-rouge">'\t'</code>) between successive values and a newline character (<code class="highlighter-rouge">'\n'</code>) at the end of each line. In order for the correlations to be read in properly, there should be <strong>no</strong> whitespace between the final correlation value and the newline character on each line.</p><p>For examples, see the <code class="highlighter-rouge">examples/</code> directory. Instructions on how to automatically generate correlations from a sequence alignment in FASTA format (and others) can be found in the Matlab file in the <code class="highlighter-rouge">scripts/</code> directory. The correlations can also be prepared through a set of Python scripts, <code class="highlighter-rouge">scripts/ACEtools.py</code>, which also includes useful auxiliary functions.</p><h1 id="running-the-program">Running the program</h1><p>Here we show a simple example of how to run the program and interpret the output, using a set of sample data for the HIV protein p7. Full explanations for the possible options are given <a href="#command-line-options">here</a>.</p><h3 id="running-ace">Running ACE</h3><p>We begin running the ACE algorithm on the example p7 dataset with the command:</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>./bin/ace <span class="nt">-d</span> examples <span class="nt">-i</span> p7 <span class="nt">-o</span> p7-out <span class="nt">-g2</span> 0.0002 <span class="nt">-b</span> 4130
</code></pre></div></div><p>This creates two new files, <code class="highlighter-rouge">examples/p7-out.sce</code> and <code class="highlighter-rouge">examples/p7-out.j</code>, which record general output on the inference procedure and the current inferred Potts parameters, respectively.</p><p>Output from the first file, <code class="highlighter-rouge">examples/p7-out.sce</code>, should appear something like the following:</p><blockquote><p>8.463532e-04 1.917385e+01 9.964734e+00 1.309919e+02 1.186601e+01 4 3321 364 8.060507e-04 1.911941e+01 1.019924e+01 1.334989e+02 1.186352e+01 4 3343 373 7.676673e-04 2.298425e+01 1.194769e+01 1.659755e+02 1.186115e+01 4 3353 380 7.311117e-04 2.550650e+01 1.277885e+01 1.768295e+02 1.185510e+01 4 3399 390 6.962969e-04 1.842427e+01 9.891297e+00 1.219612e+02 1.185371e+01 4 3442 406 …</p></blockquote><table><tbody><tr><td>These columns represent, respectively: the current value of the threshold $\theta$, error on the one-point correlations $\epsilon_{p1}$, error on the pairwise correlations $\epsilon_{p2}$, normalized maximum error $\epsilon_{\rm max}$, current estimate of the entropy $S$, maximum cluster size, total number of clusters in the expansion, and the number of selected clusters (i.e. those for which $</td><td>\Delta S</td><td>&gt;\theta$).</td></tr></tbody></table><p>The inferred Potts parameters in the second file, <code class="highlighter-rouge">examples/p7-out.j</code>, are output in the same format as the input correlations, as shown <a href="#required-input">above</a>. In this case, the first $N$ lines record the Potts fields $h_i(a)$, and the following $N(N-1)/2$ lines record the couplings $J_{ij}(a,b)$.</p><p>The final line of <code class="highlighter-rouge">examples/p7-out.sce</code> should then appear something like:</p><blockquote><p>1.338016e-05 3.641289e-01 1.674574e-01 9.498217e-01 1.169706e+01 6 20293 4773</p></blockquote><p>The error for the Potts parameters is low (the error terms $\epsilon_{p1}, \epsilon_{p2}, \epsilon_{\rm max}&lt;1$), but we can follow this initial inference step by running the <b>M</b>onte <b>C</b>arlo (MC) learning algorithm to ensure convergence. This is particularly useful when convergence is difficult to obtain in the cluster algorithm alone. Typically we find that MC learning is more likely to be successful when the entropy has nearly converged (see column 6 in <code class="highlighter-rouge">examples/p7-out.sce</code>).</p><h3 id="running-the-mc-learning-algorithm-qls">Running the MC learning algorithm QLS</h3><p>We now run the MC algorithm on the output we previously obtained from ACE, using the command:</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>./bin/qls <span class="nt">-d</span> examples <span class="nt">-c</span> p7 <span class="nt">-i</span> p7-out <span class="nt">-o</span> p7-out-learn <span class="nt">-g2</span> 0.0002 <span class="nt">-b</span> 4130
</code></pre></div></div><p>This creates two additional output files, <code class="highlighter-rouge">examples/p7-out-learn.fit</code> and <code class="highlighter-rouge">examples/p7-out-learn.j</code>, which record progress on the MC learning procedure and the current refined Potts parameters, respectively.</p><p>Output from the first file, <code class="highlighter-rouge">examples/p7-out-learn.fit</code>, should appear something like the following:</p><blockquote><p>1 1.877327e-01 1.266454e-01 1.303780e+00 1.900000e+00 2 1.976668e-01 1.284943e-01 1.329022e+00 3.610000e+00 3 1.956319e-01 1.276321e-01 1.340715e+00 6.859000e+00 4 1.959972e-01 1.282958e-01 1.279392e+00 1.303210e+01 5 2.073654e-01 1.295238e-01 1.377221e+00 2.476099e+01 …</p></blockquote><p>These columns represent, respectively: the current iteration, error on the one-point correlations $\epsilon_{p1}$, error on the pairwise correlations $\epsilon_{p2}$, normalized maximum error $\epsilon_{\rm max}$, and the maximum size of the weight parameter used in the MC learning update step. Note that the error is slightly different in this case than at the end of the cluster algorithm. This is because, by default, the MC learning algorithm computes the correlations using a larger number of samples.</p><p>After about 15 iterations the MC learning algorithm should converge and the program will terminate. The Potts parameters recorded in the second file, <code class="highlighter-rouge">examples/p7-out-learn.j</code>, now specify a model that accurately recovers the input correlations to within fluctuations expected due to finite sampling.</p><h3 id="verifying-the-output-with-qgt">Verifying the output with QGT</h3><p>If the input correlations are generated using the matlab script included in this package (for instructions, see the script itself, which lies in the <code class="highlighter-rouge">scripts/</code> directory), auxiliary measurements such as higher order correlations can be checked comprehensively. By default, this routine compares the one- and two-point correlations for the model and data, as well as the probability $P(k)$ of observing $k$ differences between sampled configurations and the “consensus” (determined from input read in to the program). It is also possible to compute the three-point correlations, the energy distribution, and a set of sample configurations.</p><p>This routine can be run using, for example:</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$ </span>./bin/qgt <span class="nt">-d</span> examples <span class="nt">-c</span> p7 <span class="nt">-m</span> p7 <span class="nt">-w</span> p7 <span class="nt">-i</span> p7-out <span class="nt">-o</span> p7-out-fit <span class="nt">-g2</span> 0.0002 <span class="nt">-b</span> 4130
</code></pre></div></div><p>This program also gives the RMS error and Pearson correlation between the model and data correlations.</p><p>Note that the output correlations from this program include not just the input ones, but also the ones corresponding to the “gauged” states, which are implicit in the input data. We also note that by default the output of small three-point correlations is trimmed in order to prevent generating extremely large files. For more information, see the options below.</p><h1 id="troubleshooting">Troubleshooting</h1><p>On difficult data sets (for example, systems of very large size, those with many states, and/or high variability), ACE may be more slow to converge. Below are a few common potential problems and suggestions for how to fix them.</p><p><strong>The one- and two-point error terms have converged ($\epsilon\leq 1$), but the maximum error remains $&gt; 1$.</strong> Obtaining a normalized maximum error $&lt; 1$ is the most stringent statistical check of the inferred model performed by the ACE and QLS routines. It is possible to obtain a very good generative model of the data even in cases where this error is larger than 1 – use QGT to verify an acceptable fit.</p><p><strong>The entropy is oscillating.</strong> Large oscillations of the entropy can be observed when whole collections of variables strongly interact. One common source is strongly-correlated gaps at the beginning and end of sequences from protein families. In such cases, filling in gaps while computing the correlations, using stronger compression, or increasing the regularization strength can help to reduce oscillations and improve convergence.</p><p><strong>ACE slows down as the size of clusters increases.</strong> This can occur if the number of states is very large. Stronger compression can allow the cluster expansion algorithm to proceed further. In addition, QLS can often converge rapidly after the network of strong interactions has been inferred by ACE, even if the errors are high and ACE has not yet advanced to a low value of the threshold (see the lattice protein model [here][1] for an example).</p><p><strong>Additional questions?</strong> Please contact us for more information or advice on dealing with difficult data sets.</p><h1 id="command-line-options">Command line options</h1><h3 id="options-for-all-programs">Options for all programs</h3><ul><li><code class="highlighter-rouge">-d</code> gives the path to the directory where data files are located, and where output will be written (default: “.” (current directory))</li><li><code class="highlighter-rouge">-i</code> specifies the name of the input file (excluding the extension, default: “input”)</li><li><code class="highlighter-rouge">-o</code> specifies the name of the output file (excluding the extension, default: “output”)</li><li><code class="highlighter-rouge">-v</code> enables verbose output</li><li><code class="highlighter-rouge">-b</code> tells the program how many samples were used to generate the input correlations, so that the expected error in the correlations due to finite sampling can be estimated (default: 1000)</li><li><code class="highlighter-rouge">-mcb</code> gives the number of Monte Carlo steps used to estimate the inference error (default: 40000 (ace), 800000 (qls, qgt))</li><li><code class="highlighter-rouge">-mcr</code> gives the number of independent Monte Carlo trajectories to use when estimating the inference error (default: 1)</li><li><code class="highlighter-rouge">-g2</code> sets the $L_2$-norm regularization strength (note that a natural value for this parameter is $1/B$, where $B$ is the number of samples used to generate the input correlations – for contact prediction, it may be best to use strong regularization $\approx 1$, regardless of the number of samples, default: 0)</li><li><code class="highlighter-rouge">-ag</code> automatically sets the $L_2$-norm regularization strength equal to $1/B$, using the number of samples $B$ passed with the <code class="highlighter-rouge">-b</code> option</li><li><code class="highlighter-rouge">-gi</code> enable the alternate gauge-invariant form of the $L_2$ regularization for couplings (see [here][1] for details)</li></ul><h3 id="additional-ace-options">Additional ACE options</h3><ul><li><code class="highlighter-rouge">-kmin</code> sets the minimum cluster size required before the program will terminate (default: 0)</li><li><code class="highlighter-rouge">-kmax</code> sets the maximum cluster size; the program terminates automatically after a cluster of this size is created (default: none)</li><li><code class="highlighter-rouge">-t</code> specifies a single value of the threshold $\theta$ at which the algorithm will run, then exit</li><li><code class="highlighter-rouge">-tmax</code> specifies the maximum (starting) value of the threshold (default: 1)</li><li><code class="highlighter-rouge">-tmin</code> specifies the minimum allowed value of the threshold; the program terminates automatically after $\theta$ falls below this minimum value (default: 1e-10)</li><li><code class="highlighter-rouge">-ts</code> specifies the logarithmic step size to for between successive values $\theta$, through $\theta_{i+1} = \theta_i / \theta_{\rm step}$ (default: 1.05)</li><li><code class="highlighter-rouge">-r</code> enables the expansion of the entropy $S$ around a mean-field reference entropy $S_{0}$, which may be helpful in particular for inferring models described by dense networks of weak interactions (note: works only if all variables are binary)</li><li><code class="highlighter-rouge">-g0</code> sets the $L_0$-norm regularization strength, and turns on $L_0$-norm regularization, enforcing sparsity for Potts couplings (default: 1e-4)</li><li><code class="highlighter-rouge">-l0</code> turns on $L_0$-norm regularization, but <strong>without</strong> setting the regularization strength</li><li><code class="highlighter-rouge">-ss</code> specifies an input “secondary structure” file used to specify the initial set of clusters to consider in the expansion</li><li><code class="highlighter-rouge">-lax</code> enables a laxer cluster construction rule, increasing the number of clusters included in the cluster expansion routine</li></ul><h3 id="additional-qls-options">Additional QLS options</h3><ul><li><code class="highlighter-rouge">-c</code> specifies the set of (true) correlations to compare with for the MC learning routine (default: “input”)</li><li><code class="highlighter-rouge">-e</code> sets the maximum tolerable error threshold; the program will run until all of the error terms $\epsilon_{p1}, \epsilon_{p2}, \epsilon_{\rm max}&lt;e$ (default: 1)</li></ul><h3 id="additional-qgt-options">Additional QGT options</h3><ul><li><code class="highlighter-rouge">-c</code> specifies the file giving the consensus sequence (default: “input”)</li><li><code class="highlighter-rouge">-m</code> specifies the file containing the compressed representation of the data (i.e. the compressed MSA, default: “input”)</li><li><code class="highlighter-rouge">-w</code> specifies the file with weights for each configuration in the data (default: “input”)</li><li><code class="highlighter-rouge">-pthresh</code> sets the threshold for three-point correlations that will be printed (default: $10\times \langle p\rangle$, where $\langle p \rangle$ is the average one-point correlation)</li><li><code class="highlighter-rouge">-p3</code> enables computation and comparison of three-point correlations; note that by default not all correlations are printed (see <code class="highlighter-rouge">-pthresh</code> option above)</li><li><code class="highlighter-rouge">-p3full</code> enables computation and comparison of three-point correlations and sets <code class="highlighter-rouge">pthresh</code> to zero</li><li><code class="highlighter-rouge">-nmax</code> set the maximum number of three-point correlations to print, another way to control file size (default: none)</li><li><code class="highlighter-rouge">-msaout</code> enables output of sample configurations from the model and their energies</li></ul><h1 id="references">References</h1><ol><li><p><a href="https://dx.doi.org/10.1093/bioinformatics/btw328">Barton, J. P., De Leonardis, E., Coucke, A. and Cocco, S. (2016). ACE: adaptive cluster expansion for maximum entropy graphical model inference. <i>Bioinformatics</i>, <b>32</b>(20): 3089-3097. DOI: 10.1093/bioinformatics/btw328</a>.</p></li><li><p><a href="https://doi.org/10.1103/PhysRevLett.106.090601">Cocco, S. and Monasson, R. (2011). Adaptive Cluster Expansion for Inferring Boltzmann Machines with Noisy Data. <i>Physical Review Letters</i>, <b>106</b>, 090601. DOI: 10.1103/PhysRevLett.106.090601</a>.</p></li><li><p><a href="https://doi.org/10.1007/s10955-012-0463-4">Cocco, S. and Monasson, R. (2012). Adaptive Cluster Expansion for the Inverse Ising Problem: Convergence, Algorithm and Tests.<i> Journal of Statistical Physics</i>, <b>147</b>(2), 252–314. DOI: 10.1007/s10955-012-0463-4</a>.</p></li><li><p><a href="https://dx.doi.org/10.1088/1742-5468/2013/03/P03002">Barton, J. and Cocco, S. (2013). Ising models for neural activity inferred via selective cluster expansion: structural and coding properties. <i>Journal of Statistical Mechanics: Theory and Experiment</i>, <b>2013</b>(03), P03002. DOI: 10.1088/1742-5468/2013/03/P03002</a>.</p></li></ol><blockquote><p>Written with <a href="https://stackedit.io/">StackEdit</a>.</p></blockquote></div></div></div></div><footer class="footer"> <a href="https://www.github.com/bartonlab" target="_blank"><i class="fa fa-github" aria-hidden="true"></i></a> <a href="https://scholar.google.com/citations?user=ItAcAOMAAAAJ" target="_blank"><i class="fa fa-book" aria-hidden="true"></i></a> <a href="https://twitter.com/_jpbarton" target="_blank"><i class="fa fa-twitter" aria-hidden="true"></i></a> <a href="mailto:john.barton@ucr.edu" target="_blank"><i class="fa fa-envelope" aria-hidden="true"></i></a><div class="post-date"><a href="/menu/about.html">Barton lab | UCR Physics</a></div> </footer></div></body></html>